### PR TITLE
Add folding and simplication for OP_ECLASS

### LINKED ISSUE #537: Extended character classes: optimisation
As a follow-up to my PR, I will refactor the parser which consumes META codes.

It should do constant-folding on simple cases, such as `[a-z -- m]` or `[^[^a-z]]`. We do still need the OP_ECLASS code, for tricky cases that can't be folded (complex logical ops over XCLASS), but doing AND/OR operations on an OP_CLASS can be folded at compile-time.

### PR DESCRIPTION
**Pull request by NWilson**
Fixes #537

Whew! This was a bit of a brain-bending PR, for me at least.

OP_ECLASS now _only_ contains ECS_XCLASS, and ECL_AND...ECL_XOR codes. No more CLASS, NCLASS nested inside; we fully-fold those, so anything like `(?[  [a-z] - [aeiou] ])` will no longer generate an OP_ECLASS. You have to actually put two XCLASSes inside the extended character class, in order to something that we can't fully fold.

In addition, for the cases that do result in an OP_ECLASS (mainly, logical operations on Unicode classes), we hoist up the bitmap so that ASCII characters can be matched without needing to consult any of the unicode properties.

### CODE REVIEW

File: HACKING
Comment (zherczeg) [inline HACKING:649]: I think you mentioned that not is removed. Is there a special type (sequence) for all/nothing?

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:792]: Maybe we should introduce static asserts.

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:6203]: If you use uint32_t type (can be a union), processing these is much faster.

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:6218]: Can we compute this at length computation time, store the bit in the META code, and use this knowledge when the byte code is generated. We might do a buffer overrun otherwise.

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:6224]: No need braces for this

File: src/pcre2_compile.h
Comment (zherczeg) [inline src/pcre2_compile.h:234]: I mentioned this could be an uint32_t array as well using an union, and merging / counting 32 bit values are 4 times faster than bytes.

File: src/pcre2_compile_class.c
Comment (zherczeg) [inline src/pcre2_compile_class.c:2409]: XNOR?

**Review by zherczeg — COMMENTED** [review]
Nice progress! I will check the rest later.

File: src/pcre2_compile_class.c
Comment (NWilson) [reply src/pcre2_compile_class.c:2409]: Yes, I think XNOR exists.

`A XOR B ⇔ (A ∧ ¬B) ∨ (¬A ∧ B)`
`A XNOR B ⇔ (A ∧ B) ∨ (¬A ∧ ¬B)`

File: HACKING
Comment (NWilson) [reply HACKING:649]: I hadn't mentioned them in the HACKING document, because they are not part of a fully-formed OP_ECLASS, but yes, there are two values ECL_ANY and ECL_NONE.

When one of these appears as the LHS/RHS of a binary operator, it can be constant-folded away. Therefore, in the final return value from `compile_class_nested()`, the result either has no instances of ECL_ANY or ECL_NONE, or the result is a single ECL_ANY or ECL_NONE item. In this case, we don't need to wrap it in an OP_ECLASS, so at match-time, an OP_ECLASS will only ever contain ECL_XCLASS items plus operators.

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:792]: This _is_ a static assert. I guess you mean, you'd prefer to wrap it in a macro? We could do that.

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6203]: Hmm. There are existing uses of `uint8_t[8]`, such as `pcre2_real_code.start_bitmap` (which has a few 32-count loops).

But that makes sense, I'll replace with a `class_bits_storage`.

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6218]: There shouldn't be a buffer overrun?

This is the rule I'm following: "don't do any conditional logic / branches inside "`if (lengthptr == NULL)`" blocks". It's totally OK to check whether lengthptr == NULL, but you need to absolutely sure when you do that, that the code will behave the same with and without lengthptr.

What would be terrible, would be to do "better" optimisation when `lengthptr == NULL`, or to do anything different at all (apart from actually writing the bytecode itself). We have to make exactly the same optimisation decisions in each case. The `if (lengthptr == NULL)` blocks need to extremely short, and careful, so that they don't affect control flow.

That way, we can be sure we'll have enough buffer space - because we followed all the same code paths each time.

**Comment by NWilson** [comment]
I know the tests are failing. I'll sort that out!

I'm also going to do additional testing.

I plan to write a grammar-driven fuzzing script, to generate millions of random extended classes which have valid syntax, and deep, complex expressions. Then I'll run pcre2test using the OLD code, to generate a reference answer for each.

If pcre2test produces the same output with the NEW code, then we can be almost certain that the optimisations don't have any effect on behaviour.

**Comment by NWilson** [comment]
Actually, _don't_ start reviewing this quite yet.

A big chunk of the core logic is all present & correct, but there are some bits of work-in-progress. The code is close, but there are some very fiddly details around XCLASS handling I'm still wrestling with.

**Comment by zherczeg** [comment]
Just a question: if multiple XCLASS-es uses property accesses, they each read the ucd data structure separately. On the long run, it would be good to do it only once, and share the ucd data pointer. A flag bit could be useful in ECLASS, which tells, that at east one XCLASS needs ucd data. Then ECLASS reads the data, and shares with XCLASS-es. Adding such flag is easy or difficult in your current implementation?

**Comment by NWilson** [comment]
Great! There are some minor bugs still to fix (CI tells me that the no-Unicode build has some failing tests).

However, is working nearly perfectly, and I don't expect any substantial changes from here on.

The code which builds XCLASSes (using char lists etc) is incredibly complex, and it would be rather nice to try hard to simplify it at some point. There are two very separate codepaths:
- The case for 16-bit/32-bit, or 8-bit with UTF set. These go via cranges, but also share code with the pure-8-bit-chars codepath, in a bit of a messy arrangement
- The pure-8-bit codepath, with no cranges, for 8-bit library without UTF set. This basically has its own special-cased code, but it's mingled in with other cases.

Fortunately... this PR really doesn't go into any of that stuff, even though it builds on top of it, so I've simply left it alone for now. However, I am very tempted to do a follow-up refactor of `PRIV(compile_class_not_nested)()`.

**Comment by zherczeg** [comment]
Let us know when this patch is ready for review.

**Comment by NWilson** [comment]
I have some additional testing to do (eg add some tests where we don't have 100% coverage of all the optimisation paths). So it's not ready to _merge_.

But the code is stable and ready for review.

File: src/pcre2_compile.c
Comment (NWilson) [inline src/pcre2_compile.c:6154]: This is a question for reviewers.

The handling for META_CLASS_EMPTY_NOT emits OP_ALLANY, and sets the flags:

>    if (firstcuflags == REQ_UNSET) firstcuflags = REQ_NONE;
>    zerofirstcu = firstcu;
>    zerofirstcuflags = firstcuflags; 

And, the handling for `[\d\D]` also emits OP_ALLANY, but sets the different flags:

>    if (firstcuflags == REQ_UNSET) firstcuflags = REQ_NONE;
>    zerofirstcu = firstcu;
>    zerofirstcuflags = firstcuflags;
>    zeroreqcu = reqcu;
>    zeroreqcuflags = reqcuflags;

Is this a problem? Is there any impact?

File: HACKING
Comment (NWilson) [inline HACKING:640]: Would someone mind checking I've understood correctly the behaviour of the bitmap for OP_XCLASS?

I made a bug in an earlier version of the PR, where I had assumed that, if an XCLASS had no bitmap, then it wouldn't match any ASCII characters. I now realise that is wrong.

If we're willing to spend 32 bytes though, we could speed up matching of things like `\pL` if we _did_ include the bitmap for all cases of XCLASS which match in the ASCII range.

**Comment by NWilson** [comment]
I have added additional tests, until the code-coverage report reached 100% branch coverage for all the (reachable) branches in the ECLASS code.

I will do further testing by fuzz-testing the new (optimised) ECLASS behaviour against the old reference implementation.

File: HACKING
Comment (zherczeg) [reply HACKING:640]: Bitmap is generated, if cranges has a range, which start is less than 256. The \p or \P properties might match to characters < 256 even if there is no bitmap.

Comment (NWilson) [reply HACKING:640]: Great! That matches my understanding now.

**Comment by NWilson** [comment]
I have completed my testing.

I wrote a script, which generates random extended character classes, UTS#18 and Perl-style, in order to fuzz-test the optimisations in this PR. I generated millions of test cases (GB of testinput), and ran the test cases using pcre2test, comparing master to this branch. The output files were identical.

Then I ran my fuzzed inputs using the Clang's coverage check... and found I'd forgotten to generate cases for one of the operators! Oops. So I re-ran... all good. The coverage verifies that the randomised test cases are covering all the branches of the optimiser, and probably covering far more paths through the code than the manually-written test cases.

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:6258]: You can move the entire `#ifndef` into the `else` below.

File: src/pcre2_compile_class.c
Comment (zherczeg) [inline src/pcre2_compile_class.c:1852]: I think this is the format:
```
switch (op)
  {
  case X:
  break;
  }
```

File: src/pcre2_compile_class.c
Comment (zherczeg) [reply src/pcre2_compile_class.c:2421]: I mean is it the combination of ECL_XOR / ECL_NOT?

File: testdata/testoutput2
Comment (zherczeg) [inline testdata/testoutput2:19975]: Nice!

File: testdata/testoutput2
Comment (zherczeg) [inline testdata/testoutput2:20141]: Is there tests with /x and comments / spaces?

File: testdata/testoutput5
Comment (zherczeg) [inline testdata/testoutput5:6241]: I think this is exactly the case where no bitmap is needed.

**Review by zherczeg — CHANGES_REQUESTED** [review]
Nice progress!

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6221]: Whenever I rewind the `code` pointer, I increment the `lengthptr` by the same amount.

* Say, we write two bytes to `code`, and request another 5 via `lengthptr`. (A total of 7 bytes requested)
* Then we rewind `code` (by 2 bytes), and increment `lengthptr` to compensate. (Still 7 bytes requested)
* Now we write more bytes to `code` - can initially decrement `lengthptr` (if writing 7 bytes or less), but then if we are requesting more bytes, we'll increase the size of the allocation

File: src/pcre2_compile_class.c
Comment (NWilson) [reply src/pcre2_compile_class.c:2421]: Yes, I have implemented XNOR by writing an XOR followed by NOT, rather than giving it a dedicated ECL_ code.

File: .github/workflows/dev.yml
Comment (NWilson) [reply .github/workflows/dev.yml:71]: Yes. The fuzz tests generate very deep stacks. The stack usage for ordinary builds is quite sensible (15 * 32 bytes), but the stack frames and stack allocations have absurdly-large amounts of guard space in the ASAN builds, so I had to bump this higher.

The eclass code (with a 15-deep parenthesis limit) uses similar or less stack than the main regex parser, which is also recursive-descent with a 255-deep limit.

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6145]: Yes. I just noticed a load of `sizeof(uint8_t)` and thought it was simpler to just remove (personal style).

File: src/pcre2_compile_class.c
Comment (NWilson) [reply src/pcre2_compile_class.c:1852]: Thanks! I'll fix these formatting issues.

File: testdata/testoutput5
Comment (NWilson) [reply testdata/testoutput5:6241]: Yes, sadly. Because ECLASS combines bitmaps in ways that are harder to reason about, I made the ECLASS bitmap behaviour a bit different to XCLASS (see the changes in HACKING). For ECLASS, I made it so that there is a bitmap if and only if the ECLASS matches any characters < 256. So an ECLASS will have at most one bitmap, but it will be preserved for cases like `\pL & \pN`.

(This example is even worse, because additional optimisation would actually mean it could be simplified to a simple XCLASS.)

Comment (zherczeg) [reply testdata/testoutput5:6241]: Ok. My idea was to detect that bitmap is needed in the `lengthptr != NULL` phase, and use this info later. I will try to do it after this patch is landed.

File: src/pcre2_compile.c
Comment (zherczeg) [reply src/pcre2_compile.c:6154]: I haven't spent time to understand these constructs, so this is also magic to me.

File: src/pcre2_compile.c
Comment (zherczeg) [inline src/pcre2_compile.c:6205]: The syntax for these is not changed.

**Review by zherczeg — APPROVED** [review]
Please squash the commits, and fix the minor things. Good progress.

**Comment by zherczeg** [comment]
Is there anybody who plans to review it later? If not, I plan to land it after a few changes.

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6205]: I did change the indentation. Is it wrong to put it all on one line?

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c:6154]: Perhaps @PhilipHazel could comment on this one specific question.

**Comment by NWilson** [comment]
Squashed. This should be good to go!

File: src/pcre2_compile.c
Comment (NWilson) [reply src/pcre2_compile.c]: I have split the code onto multiple lines.

File: src/pcre2_compile.c
Comment (PhilipHazel) [reply src/pcre2_compile.c]: Always happy to hepple. :-)

**Comment by PhilipHazel** [comment]
I tried merging this ... after a while the merge button came back so (stupidly) I clicked it again. It now says merge already in progress... has it done it, I wonder? I hope I haven't broken anything.

**Comment by zherczeg** [comment]
I still see the "squash and merge" button, but I don't click on it to avoid even more chaos.

**Comment by PhilipHazel** [comment]
I now see that too, but likewise, I'm holding back. @NWilson what do you think? Is it safe to try merging again?

**Comment by NWilson** [comment]
I can't see what buttons you're seeing... but why not? It can't hurt to keep clicking until it merges.

**Pull request merged**
